### PR TITLE
Track HTLCs in rfq policies

### PR DESCRIPTION
**Pull request by GeorgeTsagk**
## Description

This PR adds a simple in-memory tracking of accepted htlcs in the RFQ policy level. HTLCs that passed the policy check are now tracked by each policy, possibly affecting future policy check outcomes (depends on individual policy's implementation)

### CODE REVIEW

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:181]: Not sure if these need to be atomic values? Since multiple goroutines could manipulate them concurrently.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:336]: This would work for a single attempt. But in case where multiple attempts over the same edge node are done, this would quickly result in the HTLCs being rejected, because we would never reduce the current amount based on the HTLCs that were canceled from previous attempts.

**Review by guggero — CHANGES_REQUESTED** [review]
As mentioned offline, we also need to track and take into account canceled HTLCs from previous attempts.
These can be subscribed to with https://lightning.engineering/api-docs/api/lnd/router/subscribe-htlc-events.

**Review by ffranr — COMMENTED** [review]
I'm up-to-date with this PR. I agree with Oli's comments.

**Comment by GeorgeTsagk** [comment]
@ffranr @guggero 
I updated the PR to now include an htlc event subscription. 
We now maintain a map of htlc circuit keys to rfq policy
When an htlc forward is failed we retrieve the htlc circuit key, fetch the corresponding policy and have it to stop tracking the htlc.

**Comment by ffranr** [comment]
> Will also see how to add some coverage here

@GeorgeTsagk should i review this before you add coverage or wait for that extra changes?

**Comment by GeorgeTsagk** [comment]
> > Will also see how to add some coverage here
> 
> @GeorgeTsagk should i review this before you add coverage or wait for that extra changes?

You can review
Coverage will soon be added in the LitD itests

File: tapchannel/aux_leaf_signer.go
Comment (ffranr) [reply tapchannel/aux_leaf_signer.go:772]: Also, on the function name, perhaps `ScriptKeyTweakFromHtlcIndex`.

File: tapchannel/aux_leaf_signer.go
Comment (GeorgeTsagk) [reply tapchannel/aux_leaf_signer.go:787]: this was from a diff PR this was based on, ignore

File: tapchannel/aux_leaf_signer_test.go
Comment (GeorgeTsagk) [reply tapchannel/aux_leaf_signer_test.go:449]: this was from a diff PR this was based on, ignore

File: tapchannel/aux_leaf_signer.go
Comment (GeorgeTsagk) [reply tapchannel/aux_leaf_signer.go:796]: this was from a diff PR this was based on, ignore

File: chain_bridge.go
Comment (guggero) [inline chain_bridge.go:380]: nit: requires Godoc comment (for the same method in the `HtlcSubscriber` interface as well).

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:115]: So both the `htlcToAmt` and `CurrentAmountMsat` need to be in sync. If we use an atomic value for one and a sync map for the other, then it's still possible for a _read_ from another goroutine to get a different value while we're updating things (e.g. the map was updated but the total not yet).

So I think because we have two separate values, we need a `sync.RWMutex` instead. But then we can use a normal map here at least.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:172]: nit: commented out code.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:396]: I think we can probably simplify the interface of this `TrackAcceptedHtlc` a bit by just passing in `circuitKey` and `amount`. Then it's IMO a bit more apparent what's happening (looks more like a map add and delete) and the policy doesn't need to know how the key is calculated.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:192]: We can use structs as map keys directly. And as long as there are no pointer values in a struct (meaning the struct can be directly compared with `==` without the pointers leading to side effects), things work out as expected. So we can just use the circuit key directly as the map key, no need to convert to string.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:207]: Yeah, we definitely need a mutex here. Otherwise another goroutine might read this value while we're waiting on the `LoadAndDelete` above to complete.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:107]: nit: use the same capitalization for HTLC everywhere? So `s/htlcs/HTLCs` here and throughout the file.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:633]: Same here, can use the circuit key as the map key directly. Here a sync map is enough and we don't need a mutex, as it's just a single map that needs to be synced.

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:790]: nit: format as:

```
				policy, found := h.htlcToPolicy.LoadAndDelete(
					htlcIDStr,
				)
```
instead.

**Review by guggero — COMMENTED** [review]
Looks pretty good. Left a couple of comments to simplify the code and make it more concurrency safe.

File: rfq/order.go
Comment (GeorgeTsagk) [reply rfq/order.go:115]: > So both the htlcToAmt and CurrentAmountMsat need to be in sync

What do you mean by "in sync"? The `htlcToAmt` is only used to cache the htlc's effective amount when the time to "cancel" it comes (i.e call `UntrackHtlc`)

On the other hand, `CurrentAmountMsat` is incremented/decremented by multiple goroutines, so def agree that we have to guard that.

So could just use a mutex that protects `CurrentAmountMsat` alone

**Review by ffranr — COMMENTED** [review]
I'm up to date on this. Will wait for the revised commits based on Oli's review.

**Comment by GeorgeTsagk** [comment]
Thank you @ffranr @guggero 

I just pushed the commits that address your feedback, also rebased on latest #1224

File: rfq/order.go
Comment (guggero) [inline rfq/order.go:724]: This is already done by `TrackAcceptedHtlc` below.

File: rfq/order.go
Comment (GeorgeTsagk) [reply rfq/order.go:204]: oopsie

File: rfq/order.go
Comment (GeorgeTsagk) [reply rfq/order.go:724]: that's a diff map called `htlcToAmt`

File: rfq/order.go
Comment (GeorgeTsagk) [reply rfq/order.go:812]: dif map called `htlcToAmt`

File: rfq/order.go
Comment (GeorgeTsagk) [reply rfq/order.go:542]: Not necessary

Every policy needs to be atomic w.r.t it's own state, if we have two callers in the composite `TrackAcceptedHtlc` above it doesn't really matter if you guarantee that in&out policy will be updated atomically. If you have 2 HTLCs being intercepted at the same time (while only one can make it through) then you're just not sure which one is gonna get cancelled, which is the case anyway

File: rfq/order.go
Comment (guggero) [reply rfq/order.go:729]: Ah yes, my bad.

**Review by guggero — COMMENTED** [review]
LGTM pending the dependent PR.

**Comment by GeorgeTsagk** [comment]
Rebased on `main` after merging dependent PR, PTAL

**Review by guggero — APPROVED** [review]
LGTM :tada:

**Review by ffranr — COMMENTED** [review]
I'm not sure if we need
```
htlcToPolicy lnutils.SyncMap[models.CircuitKey, Policy]
```
Please see review comments.

File: rfq/order.go
Comment (ffranr) [reply rfq/order.go:668]: If we can manage to do that then all the HTLC states can be nested under the relevant policies.

Comment (GeorgeTsagk) [reply rfq/order.go:668]: Yes `OrderHandler.fetchPolicy` does find the scid but that's not done based on the `CircuitKey` alone, it also uses the `InterceptedHtlc.InWireCustomRecords` field to extract the scid directly. Unfortunately this field is not available when you get notified for the HTLC cancel event*, so that's why I'm caching this info on this layer.

\*It's not a value that's simply not exposed on the API, but more like data that is not available in the lnd channel link by the time we process the HTLC cancel. To change LND and have that data available by the API directly would involve doing the same kind of HTLC-caching on the LND link level (or something similar), which could get much more complicated.

I def agree that eliminating HTLC related state here would be a win, but can't see a workaround.

File: rfq/order.go
Comment (GeorgeTsagk) [reply rfq/order.go:400]: Sure, performance overhead is too small to battle readability

File: rfq/order.go
Comment (ffranr) [reply rfq/order.go:668]: I'm trying to remember how this work. Are there HTLCs where we can only get the SCID from their custom records? Is the custom records check strictly necessary for some HTLCs?

Comment (ffranr) [reply rfq/order.go:668]: For incoming asset payment HTLCs we inspect custom records. From what I understand and according to @guggero .

So I will approve this PR as is. I will leave merging down to you @GeorgeTsagk . I'll put up a PR to clarify `OrderHandler.fetchPolicy` with comments.

This PR should work as it is either way IMO.

**Review by ffranr — APPROVED** [review]
Will leave merging down to @GeorgeTsagk .

File: rfq/order.go
Comment (GeorgeTsagk) [reply rfq/order.go:668]: Yes, we can only retrieve the SCID from the custom records. The ChanID of the circuit keys correspond to the actual scids of the channels that carry the HTLCs

**Pull request merged**
